### PR TITLE
update translator service to bclconvert manager

### PR DESCRIPTION
**Pull request by raylrui**
## Update translator service to bclconvert manager

As discussed, 
 1. translator service will be included in BclConvert Manager.
 2. event scheme from bcm to bclconvertmanager
 3. modify in event schemas
 4. add SampleSheet Output, RunID and BaseSpace ID #301 

new schema example 
 ``` json5 
{
  "version": "0",
  "id": "f71aaaaa-5b36-40c2-f7dc-804ca6270cd6",
  "detail-type": "WorkflowRunStateChange",
  "source": "orcabus.bclconvertmanager",
  "account": "123456789012",
  "time": "2024-05-01T09:25:44Z",
  "region": "ap-southeast-2",
  "resources": [],
  "detail": {
    "portalRunId": "202405012397actg",
    "timestamp": "2024-05-01T09:25:44Z",
    "status": "SUCCEEDED",
    "workflowName": "BclConvert",
    "workflowVersion": "4.2.7",
    "workflowRunName": "540424_A01001_0193_BBBBMMDRX5_c754de_bd822f",
    "payload": {
      "refId": null,
      "version": "0.1.0",
      "data": {
        "projectId": "bxxxxxxxx-dxxx-4xxxx-adcc-xxxxxxxxx",
        "analysisId": "aaaaafe8-238c-4200-b632-d5dd8c8db94a",
        "userReference": "540424_A01001_0193_BBBBMMDRX5_c754de_bd822f",
        "timeCreated": "2024-05-01T10:11:35Z",
        "timeModified": "2024-05-01T11:24:29Z",
        "pipelineId": "bfffffff-cb27-4dfa-846e-acd6eb081aca",
        "pipelineCode": "BclConvert v4_2_7",
        "pipelineDescription": "This is an autolaunch BclConvert pipeline for use by the metaworkflow",
        "pipelineUrn": "urn:ilmn:ica:pipeline:bfffffff-cb27-4dfa-846e-acd6eb081aca#BclConvert_v4_2_7",
        "instrumentRunId": "12345_A12345_1234_ABCDE12345",
        "basespaceRunId": "1234567",
        "samplesheetB64gz": "H4sIAFGBVWYC/9VaUW+jOBD+Kyvu9VqBgST0njhWh046..."
      }
    }
  }
}
```

### CODE REVIEW

**Comment by victorskl** [comment]
Morning Ray. Great start. 

Could you do me favour; can you create a github issue/ticket with the same description - and, label "pipeline" "refactor". 

Then, pls link it to; this PR > Development > Link issue from this repository (at the right panel)

Thank so much.

File: docs/event-schemas/bcm/WorkflowRunStateChange.json [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+            "nullable": true
+          },
+          "version": {
+            "type": "string"
+          },
+          "data": {
+            "$ref": "#/components/schemas/Data"
+          }
+        },
+        "additionalProperties": true
```
Comment (victorskl) [inline docs/event-schemas/bcm/WorkflowRunStateChange.json:105]: You can remove this `"additionalProperties": true` here.

File: docs/event-schemas/bcm/WorkflowRunStateChange.json
Comment (victorskl) [inline docs/event-schemas/bcm/WorkflowRunStateChange.json:1]: Also could rename folder to

```
docs/event-schemas/bclconvertmanager/WorkflowRunStateChange.json
```

Comment (victorskl) [reply docs/event-schemas/bcm/WorkflowRunStateChange.json:1]: Related #307

File: docs/event-schemas/bcm/WorkflowRunStateChange.json (hunk 1 above)
Comment (victorskl) [reply docs/event-schemas/bcm/WorkflowRunStateChange.json:105]: Related #308

**Comment by raylrui** [comment]
Hi @reisingerf @victorskl @alexiswl ;
update for bclconvert manager:

1. create BclConvertManager stack which include translator service.
2.  add "instrumentRunId", "basespaceRunId", "samplesheetB64gz" in the required field of  `orcabus.bclconvertmanager@PayloadDataSucceeded` according #301 
3. The event tranlator then returns the following:

(Non-SUCCEEDED Event without payload)
```json5
{
    "portalRunId": '20xxxxxxxxxx',
    "executionId": "valid_payload_id",
    "timestamp": "2024-00-25T00:07:00Z",
    "status": "SUCCEEDED",
    "workflowName": "BclConvert",
    "workflowVersion": "4.2.7",
    workflowRunName: "123456_A1234_0000_TestingPattern",
}
```
(SUCCEEDED Event)
```json5
{
    "portalRunId": '20xxxxxxxxxx',
    "executionId": "valid_payload_id",
    "timestamp": "2024-00-25T00:07:00Z",
    "status": "SUCCEEDED",
    "workflowName": "BclConvert",
    "workflowVersion": "4.2.7",
    workflowRunName: "123456_A1234_0000_TestingPattern",
    "payload": {
        "refId": None,
        "version": "0.1.0",
        "data": {
            "projectId": "valid_project_id",
            "analysisId": "valid_payload_id",
            "userReference": "123456_A1234_0000_TestingPattern",
            "timeCreated": "2024-01-01T00:11:35Z",
            "timeModified": "2024-01-01T01:24:29Z",
            "pipelineId": "valid_pipeline_id",
            "pipelineCode": "BclConvert v0_0_0",
            "pipelineDescription": "BclConvert pipeline.",
            "pipelineUrn": "urn:ilmn:ica:pipeline:123456-abcd-efghi-1234-acdefg1234a#BclConvert_v0_0_0"
            "instrumentRunId": "12345_A12345_1234_ABCDE12345",
            "basespaceRunId": "1234567",
            "samplesheetB64gz": "H4sIAFGBVWYC/9VaUW+jOBD+Kyvu9VqBgST0njhWh046..."
        }
    }
}
```

File: lib/workload/stateless/stacks/bclconvert-manager/translator_service/icav2_event_translator.py [hunk 2]
```diff
@@ -0,0 +1,504 @@
...
+        detail_type="WorkflowRunStateChange",
+        source="orcabus.bclconvertmanager",
+        # version="0.1.0",  # comment as the version is managed by the evnet bus
+    )
+
+
+# Convert from entity module to internal event details
+def get_event_details(event) -> WorkflowRunStateChange:
+    # Extract relevant fields from the event payload
+    analysis_status = event.get("eventParameters", {}).get("analysisStatus", '')
```
Comment (reisingerf) [inline lib/workload/stateless/stacks/bclconvert-manager/translator_service/icav2_event_translator.py:209]: Would "UNSPECIFIED" be a better default?

File: lib/workload/stateless/stacks/bclconvert-manager/translator_service/icav2_event_translator.py [hunk 3]
```diff
@@ -0,0 +1,504 @@
...
+
+    payload = event.get("payload", {})
+    analysis_id = payload.get("id", '')
+    pipeline = payload.get("pipeline", {})
+
+    event_name, version = parse_event_code(pipeline.get("code", ''))
+    
+    if analysis_status != "SUCCEEDED":
```
Comment (reisingerf) [inline lib/workload/stateless/stacks/bclconvert-manager/translator_service/icav2_event_translator.py:217]: An empty payload for all non-SUCCEEDED events is fine for now.
In the future we may want to handle all known statuses, for example add an error message payload for FAILED events. Happy to put that on the backlog though.

**Review by reisingerf — APPROVED** [review]
LGTM 
Minor comments.

File: lib/workload/stateless/stacks/bclconvert-manager/translator_service/icav2_event_translator.py [hunk 4]
```diff
@@ -0,0 +1,504 @@
...
+    try:
+        db_uuid = generate_db_uuid(None, None).get("db_uuid",'')
+        dynamodb.put_item(
+            TableName=table_name,
+            Item={
+                'id': {'S': db_uuid},
+                'id_type': {'S': 'db_uuid'},
+                'analysis_id': {'S': internal_ica_event.detail.payload.get('data', '').get("analysisId", '')},
+                'analysis_status': {'S': internal_ica_event.detail.status},
+                # 'SUCCEEDED', 'FAILED', 'INPROGRESS', 'ABORTED', 'UNKNOWN
```
[the remark this answers is not in the document]
Comment (ray-lius) [reply lib/workload/stateless/stacks/bclconvert-manager/translator_service/icav2_event_translator.py:159]: sure, will just list the 3 final state 'SUCCEEDED', 'FAILED', 'ABORTED'.

File: lib/workload/stateless/stacks/bclconvert-manager/translator_service/icav2_event_translator.py (hunk 3 above)
Comment (raylrui) [reply lib/workload/stateless/stacks/bclconvert-manager/translator_service/icav2_event_translator.py:216]: sure, new backlog added [here](https://github.com/orgs/umccr/projects/21/views/1?pane=issue&itemId=65106537)
I think for non-final status "Requested", "Queued", "Initializing", "Preparing Inputs", "In Progress", "Generating outputs", "Aborting", we can leave it empty;
for final status "Aborted", "Failed", "Succeeded", we may need add additional specific information in the future.

File: lib/workload/stateless/stacks/bclconvert-manager/translator_service/icav2_event_translator.py (hunk 2 above)
Comment (raylrui) [reply lib/workload/stateless/stacks/bclconvert-manager/translator_service/icav2_event_translator.py]: sure, added

File: lib/workload/stateless/stacks/bclconvert-manager/translator_service/icav2_event_translator.py [hunk 5]
```diff
@@ -0,0 +1,504 @@
...
+{
+    "portalRunId": '20xxxxxxxxxx',
+    "executionId": "valid_payload_id",
+    "timestamp": "2024-00-25T00:07:00Z",
+    "status": "SUCCEEDED",
+    "workflowName": "BclConvert",
+    "workflowVersion": "4.2.7",
+    workflowRunName: "123456_A1234_0000_TestingPattern",
+    "payload": {
+        "refId": None,
```
[the remark this answers is not in the document]
Comment (raylrui) [reply lib/workload/stateless/stacks/bclconvert-manager/translator_service/icav2_event_translator.py]: refId deleted as not required

File: lib/workload/stateless/stacks/bclconvert-manager/translator_service/icav2_event_translator.py [hunk 6]
```diff
@@ -0,0 +1,504 @@
...
+    ....
+  }
+}
+The event tranlator then returns the following:
+(Non-SUCCEEDED Event without payload)
+{
+    "portalRunId": '20xxxxxxxxxx',
+    "executionId": "valid_payload_id",
+    "timestamp": "2024-00-25T00:07:00Z",
+    "status": "SUCCEEDED",
```
[the remark this answers is not in the document]
Comment (raylrui) [reply lib/workload/stateless/stacks/bclconvert-manager/translator_service/icav2_event_translator.py]: fixed

File: lib/workload/stateless/stacks/bclconvert-manager/translator_service/icav2_event_translator.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (raylrui) [reply lib/workload/stateless/stacks/bclconvert-manager/translator_service/icav2_event_translator.py]: this may come from old draft, already deleted.

**Review by victorskl — APPROVED** [review]
Couple of naming convention spotted but bypassing those for now. Looking good otherwise.

File: docs/schemas/data/bclconvertmanager/PayloadDataSucceeded.json
[the remark this answers is not in the document]
Comment (raylrui) [reply docs/schemas/data/bclconvertmanager/PayloadDataSucceeded.json:1]: sure, will deleted the deprecated `PayloadDataSucceeded` file.

**Pull request merged**